### ISSUE TITLE
Godot.RC3 .. RC5: breaks `timer.connect()`, the program is stopped forever

### ISSUE THREAD

**Issue by MikeSchulze** [godotengine/godot#73889]
### Godot version

v4.0.rc5.official [6296b4600]

### System information

Windows 19

### Issue description

After upgration to Godot.RC3 .. RC4 .. RC5 my plugin is broken.

I provide a function to wait for a function to provide an expected value.

As example it waits for function 'obj_value' from the instance `TestIterativeValueProvider` is providing null.
```
func test_foo(timeout = 20000) -> void:
	var value_provider := TestIterativeValueProvider.new(RefCounted.new(), 5, null)
	await assert_func(value_provider, "obj_value").is_null()
````
Now the execution stops at await and is not executed further

I have reduced the code as much as possible to find the root of the cause.
The problem occurs when using `timer.connect("timeout", _on_timeout)`.

### Steps to reproduce

run the attached example 

a) if you run `await assert_that().it_works()` the function is continue after the signal `done` is emitted
b) if you run `await assert_that().not_works()` the function is stuck
 --  b1) also strange when using variant (b) and remove the return type from the function `assert_that` it works too

### Minimal reproduction project

[timer_await.zip](https://github.com/godotengine/godot/files/10828504/timer_await.zip)

**Comment by dalexeev** [comment]
If you're using signal/async function chaining, then it's possible that some initial or intermediate object doesn't already exist (was freed). Are you sure the awaited signal is emitted (the awaited function is completed)?

**Comment by MikeSchulze** [comment]
yes i have analysed in deep, see the example project

**Comment by MikeSchulze** [comment]
it breaks many places in my plugin
e.g.
```
        # works
	timeout.connect("timeout", func on_timeout():_interrupted = true)
        # not works
	#timeout.connect("timeout", _on_timeout)

func _on_timeout():
	_interrupted = true


```

**Comment by dalexeev** [comment]
I tested the MRP. First, there are errors (I fixed them):

![](https://user-images.githubusercontent.com/47700418/221360527-31b21394-fa76-4a8a-b7dd-f0773e9759d5.png)

Secondly, as I assumed, the awaiting object is already freed at the moment when the signal is emitted:

![](https://user-images.githubusercontent.com/47700418/221360673-7ea376d0-e6c8-419f-8982-978a829e877c.png)

This is not yet properly documented, but there is a difference between a lambda and `_on_timeout` (read as `self._on_timeout`). In the second case, you create a `Callable` bound to an object instance. And the moment the instance is deleted, this `Callable` becomes invalid (now it's just correctly cleared after #65910). Unlike `_on_timeout`, a lambda is not bound to an object instance (in the sense that it can outlive it).

[timer_await_2.zip](https://github.com/godotengine/godot/files/10831311/timer_await_2.zip)

**Comment by dalexeev** [comment]
> also strange when using variant (b) and remove the return type from the function `assert_that` it works too

It's probably the same bug that in #70637.

**Comment by anvilfolk** [comment]
Really nice detective work @dalexeev ❤️ 

I am struggling to see the implications for this issue - does that mean that everything is working as expected, and it's more of a documentation issue? Or is there something here that needs to be fixed?

**Comment by dalexeev** [comment]
I'm not sure there are no other bugs left. Perhaps we `unreference()` a temporary object earlier than we should, or there is some other bug in the compiler/VM. This is just my intuition, I don't know enough about it.

The MRP code is hard to analyze because it has `await`, `signals`, `call_deferred`, plus we need to take into account the lifetime of objects. Personally, I would avoid code like this, but we need to understand it in order to know if GDScript works correctly.

**Comment by MikeSchulze** [comment]
@dalexeev thanks for your analyse

What i not understand you say `And the moment the instance is deleted, this Callable becomes invalid `
But the instance is not deleted, it stays at await and will not be freed before the function is left and quit is called,

**Comment by kleonc** [comment]
> This is not yet properly documented, but there is a difference between a lambda and `_on_timeout` (read this as `self._on_timeout`). In the second case, you create a `Callable` bound to an object instance. And the moment the instance is deleted, this `Callable` becomes invalid (now it's just correctly cleared after #65910). Unlike `_on_timeout`, a lambda is not bound to an object instance (in the sense that it can outlive it).
> 
> Here you can see that the "method reference" and the lambda are bound to different objects (the instance and the class):
> 
> ![](https://user-images.githubusercontent.com/47700418/221361276-a87bd418-a56e-487d-b70f-e6b69c810670.png)

@dalexeev But seems that if the lambda somehow references the instance then it's bound to the instance instead of to the script:
```gdscript
var f3 = func(): self
print(f3.get_object() == self) # true
```

That's what seems to happen in the MRP, the lambda keeps the `RefCounted` object (`AssertThat`) alive.

---

I think this showcases the main issue in here:
```gdscript
extends Node

func _ready() -> void:
	print(1)
	await A.new().some_signal
	print(2)

class A:
	signal some_signal

	func _init() -> void:
		call_deferred("emit_signal", "some_signal")
```


- In v4.0.rc2.official [d2699dc7a] outputs:

```
1
2
```
and produces errors:
```
E 0:00:00:0730   ~Object: Object <Object#-9223372013635108439> was freed or unreferenced while a signal is being emitted from it. Try connecting to the signal using 'CONNECT_DEFERRED' flag, or use queue_free() to free the object (if this object is a Node) to avoid this error and potential crashes.
  <C++ Source>   core/object/object.cpp:1783 @ ~Object()
E 0:00:00:0730   _disconnect: Disconnecting nonexistent signal 'some_signal' in <Object#0>.
  <C++ Error>    Condition "!s" is true.
  <C++ Source>   core/object/object.cpp:1329 @ _disconnect()
```


- In v4.0.rc3.official [7e79aead9], v4.0.rc5.official [6296b4600] outputs:

```
1
```
and breaks at the await line with an error:
```
Error connecting to signal: some_signal during await.
```
```
E 0:00:00:0710   test2.gd:4 @ _ready(): Condition "!obj" is true. Returning: ERR_UNCONFIGURED
  <C++ Source>   core/variant/callable.cpp:459 @ connect()
  <Stack Trace>  test2.gd:4 @ _ready()
```
After resuming the execution `2` is not printed.

So seems like the just created `RefCounted` object is being freed before `await` connects to its signal. Storing such object in a variable works as a workaround:

```gdscript
extends Node

func _ready() -> void:
	print(1)
	var obj = A.new()
	await obj.some_signal
	print(2)

class A:
	signal some_signal

	func _init() -> void:
		call_deferred("emit_signal", "some_signal")
```
This works in v4.0.rc5.official [6296b4600] the same as the `await A.new().some_signal` version works in v4.0.rc2.official [d2699dc7a] (outputs `1`, `2` and errors about disconnecting).

**Comment by dalexeev** [comment]
@kleonc Then I guess that the reason is #73448. It is merged into RC 3.

> But the instance is not deleted, it stays at await and will not be freed before the function is left and quit is called.

The `AssertThat` instance is freed before the timer signal is emitted, not the awaiting node and not the timer.

> An question came up, if we connect a lambda where is do the disconnect?
> Is this still done when the timer is freed?

When an object is freed, all connections to it are removed as well.

https://github.com/godotengine/godot/blob/3863199ab940272f6844ff30910ec7a520e47f41/core/object/object.cpp#L1802-L1806

**Comment by MikeSchulze** [comment]
> The AssertThat instance is freed before the timer signal is emitted, not the awaiting node and not the timer.

Thats what i not understand,  the instance should be live antil the while loop and at least it is returning the instance it self to call `_runB()`
In my point of view the instance should only freed after left function `.not_works()`

**Comment by kleonc** [comment]
> Then I guess that the reason is #73448. It is merged into RC 3.

cc @RandomShaper

**Comment by dalexeev** [comment]
> In my point of view the instance should only freed after left function `.not_works()`

Yes, this appears to be a bug. But it is also important for us to avoid the opposite problem: a memory leak due to a dangling reference in the case `await` does not complete for one reason or another.

**Comment by vnen** [comment]
Great detective work!

So, in hindsight I can see how freeing the temporaries would be problematic. It is only safe to free them after the end of a statement, since they are not meant to live across statements anyway.

Though in the current state they are only freed after being used, at least in theory, which should be okay. It might need more investigation.

Unfortunately I think this bug will stay in 4.0 and will have to be fixed later, as I feel like there won't be a simple fix (that we would be sure to not introduce other bugs) and reverting https://github.com/godotengine/godot/pull/73448 would re-introduce other bugs.
